test/performance/odp_dma_perf.c

```diff
@@ -280,804 +526,959 @@ static int parse_options(int argc, char **argv, test_config_t *config)
+}
 
-			++k;
-			last_ev_idx = i;
-			has_events = 1;
-		}
-	}
+static void wait_compl_event(sd_t *sd, stats_t *stats)
+{
+	uint64_t start_cc, end_cc, wait_cc, trs_tm, trs_cc, start_tm, start_cc_diff;
+	odp_event_t ev;
+	odp_dma_result_t res;
+	trs_info_t *info;
+	int ret;
 
-	last_ev_idx = has_events ? last_ev_idx + 1 : 0;
+	start_cc = odp_cpu_cycles();
+	ev = odp_schedule(NULL, odp_schedule_wait_time(ODP_TIME_SEC_IN_NS));
+	end_cc = odp_cpu_cycles();
 
-	for (int i = last_ev_idx; i < config->dma_rounds; ++i) {
-		compl_wait_entry_t entry = { .type = ODP_DMA_COMPL_POLL,
-					     .id = compl_params[i].transfer_id };
-		list[i] = entry;
+	if (odp_unlikely(ev == ODP_EVENT_INVALID)) {
+		++stats->scheduler_timeouts;
+		return;
+	}
+
+	odp_dma_compl_result(odp_dma_compl_from_event(ev), &res);
+	info = res.user_ptr;
+	trs_tm = odp_time_global_ns() - info->trs_start_tm;
+	stats->max_trs_tm = MAX(trs_tm, stats->max_trs_tm);
+	stats->min_trs_tm = MIN(trs_tm, stats->min_trs_tm);
+	stats->trs_tm += trs_tm;
+	trs_cc = odp_cpu_cycles_diff(odp_cpu_cycles(), info->trs_start_cc);
+	stats->max_trs_cc = MAX(trs_cc, stats->max_trs_cc);
+	stats->min_trs_cc = MIN(trs_cc, stats->min_trs_cc);
+	stats->trs_cc += trs_cc;
+	++stats->trs_cnt;
+	wait_cc = odp_cpu_cycles_diff(end_cc, start_cc);
+	stats->max_wait_cc = MAX(wait_cc, stats->max_wait_cc);
+	stats->min_wait_cc = MIN(wait_cc, stats->min_wait_cc);
+	stats->wait_cc += wait_cc;
+	++stats->wait_cnt;
+
+	if (odp_unlikely(!res.success))
+		++stats->transfer_errs;
+	else
+		++stats->completed;
+
+	start_tm = odp_time_global_ns();
+	start_cc = odp_cpu_cycles();
+	ret = odp_dma_transfer_start(sd->dma.handle, &info->trs_param, &info->compl_param);
+	end_cc = odp_cpu_cycles();
+
+	if (odp_unlikely(ret <= 0)) {
+		++stats->start_errs;
+	} else {
+		info->trs_start_tm = start_tm;
+		info->trs_start_cc = start_cc;
+		start_cc_diff = odp_cpu_cycles_diff(end_cc, start_cc);
+		stats->max_start_cc = MAX(start_cc_diff, stats->max_start_cc);
+		stats->min_start_cc = MIN(start_cc_diff, stats->min_start_cc);
+		stats->start_cc += start_cc_diff;
+		++stats->start_cnt;
 	}
 }
 
-static inline int wait_dma_transfers_ready(test_config_t *config, compl_wait_entry_t list[])
+static void drain_compl_events(void)
 {
 	odp_event_t ev;
-	const uint64_t wait_time = odp_schedule_wait_time(config->wait_ns);
-	uint64_t start, end;
-	int done = 0;
 
-	for (int i = 0; i < config->dma_rounds; ++i) {
-		if (list[i].type == ODP_DMA_COMPL_EVENT) {
-			ev = odp_schedule(NULL, wait_time);
+	while (true) {
+		ev = odp_schedule(NULL, odp_schedule_wait_time(ODP_TIME_SEC_IN_NS));
 
-			if (ev == ODP_EVENT_INVALID) {
-				ODPH_ERR("Error waiting event completion.\n");
-				return -1;
-			}
-		} else {
-			start = odp_time_local_ns();
-			end = start + ODP_TIME_SEC_IN_NS;
-
-			while (1) {
-				done = odp_dma_transfer_done(config->dma_config.handle, list[i].id,
-							     NULL);
-
-				if (done > 0)
-					break;
+		if (ev == ODP_EVENT_INVALID)
+			break;
```

Comment:
Maybe add a comment here to explain that events are freed later on in `teardown_data()`. Currently this looks like events are leaking.

Reply:
It's a common pattern throughout the tester, so no comments here, but documented the test api and added a comment in `main()` in v3.